drake/core/Gradient.h

```diff
@@ -0,0 +1,57 @@
+#ifndef DRAKE_GRADIENT_H
+#define DRAKE_GRADIENT_H
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include <unsupported/Eigen/AutoDiff>
+
+namespace Drake {
+  // todo: recursive template to get arbitrary gradient order
+
+  // note: tried using template default values (e.g. Eigen::Dynamic), but they didn't seem to work on my mac clang
+  template <int num_vars> using TaylorVard = Eigen::AutoDiffScalar< Eigen::Matrix<double,num_vars,1> >;
+  template <int num_vars, int rows> using TaylorVecd = Eigen::Matrix< TaylorVard<num_vars>, rows, 1>;
+  template <int num_vars, int rows, int cols> using TaylorMatd = Eigen::Matrix< TaylorVard<num_vars>, rows, cols>;
+
+  typedef TaylorVard<Eigen::Dynamic> TaylorVarXd;
+  typedef TaylorVecd<Eigen::Dynamic,Eigen::Dynamic> TaylorVecXd;
+  typedef TaylorMatd<Eigen::Dynamic,Eigen::Dynamic,Eigen::Dynamic> TaylorMatXd;
+
+  /** initTaylorVecXd
+   * @brief initializes the vector with x=val and dx=eye(numel(val))
+   */
+  template <typename DerivedA, typename DerivedB>
+  void initTaylorVecXd(const Eigen::MatrixBase<DerivedA>& val, Eigen::MatrixBase<DerivedB>& auto_diff_vec) {
+    Eigen::MatrixXd der = Eigen::MatrixXd::Identity(val.rows(),val.rows());
```

Comment:
I think it would also be better to infer the scalar type of the derivatives from auto_diff_vec instead of assuming it's double (by creating a MatrixXd). That would make the function naturally support higher order derivatives.